controllers/validators/validators.go

```diff
@@ -0,0 +1,38 @@
+package validators
+
+import (
+	"fmt"
+
+	"github.com/blang/semver/v4"
+
+	operatorsv1alpha1 "github.com/operator-framework/operator-controller/api/v1alpha1"
+)
+
+type operatorCRValidatorFunc func(operator *operatorsv1alpha1.Operator) error
+
+// validateSemver validates that the operator's version is a valid SemVer.
+// this validation should already be happening at the CRD level. But, it depends
+// on a regex that could possibly fail to validate a valid SemVer. This is added as an
+// extra measure to ensure a valid spec before the CR is processed for resolution
+func validateSemver(operator *operatorsv1alpha1.Operator) error {
+	if operator.Spec.Version == "" {
+		return nil
+	}
+	if _, err := semver.Parse(operator.Spec.Version); err != nil {
+		return fmt.Errorf("invalid .spec.version: %w", err)
+	}
+	return nil
+}
+
+// ValidateOperatorSpec validates the operator spec, e.g. ensuring that .spec.version, if provided, is a valid SemVer
+func ValidateOperatorSpec(operator *operatorsv1alpha1.Operator) error {
+	validators := []operatorCRValidatorFunc{
+		validateSemver,
+	}
+	for _, validator := range validators {
+		if err := validator(operator); err != nil {
+			return err
```

Comment:
For now it is ok as there is only one field which requires validation. Just curious how you see UX for it once we have more validators - Will we bereturning one error at a time or use some sort of aggregated error (so we surface all issues at once)?

Reply:
That's a great call out. Personally, I think we should try as much as possible to add all of the validation errors as conditions. It would suck to keep going back and forth trying to get it right. What I'm not sure of right now is if it's "acceptable" to add multiple conditions of the same type (I don't see why not, but I don't know what the best practice is here). According to my good friend ChatGPT:

```
It is generally best to keep the status of a Kubernetes resource concise and clear. If you have multiple conditions of the same type, it might be better to consolidate them into a single condition that provides an accurate representation of the overall state of the resource.

Here are some guidelines to consider when working with conditions in the status of a Kubernetes resource:

Each condition should have a distinct type that describes the aspect of the resource being observed.
Conditions should be used to provide insights into the state of the resource and any issues that might be affecting it.
Use the Reason and Message fields within a condition to provide more detailed information about the state of the resource and any issues.
Update conditions as the state of the resource changes, ensuring that the status stays up-to-date and accurate.
By following these guidelines, you can maintain a clear and concise status that communicates the state of the resource effectively. In cases where you need to represent multiple aspects of the resource's state, it's better to use separate conditions with different types. However, if you have multiple instances of the same type of condition, consider consolidating them into a single condition with a clear reason and message.
```
It seems to suggest we should try to group everything into a single condition. But that seems ugly at first sight. Another option from the top of my head might be to create a condition for each validation error. That also seems ugly. Personally, I'm in favor of just having multiple conditions of the same type each describing its own reason for failure.

I suggest we leave it as it is for now. I will add a comment and a ticket so we can follow up on this after a more thorough discussion upstream. wdyt?

Reply:
I have definitevely seen conditions like `ValidSomething` true/false with message containing concatenated errors. I think it is very common in OCP. But I also seen condition per issue approach.

IMO - if condition is meant to be consumed by humans - we might just want to concatenate errors. We will later be able to split them into separate conditions. I personally not a fan of scrolling through 10s of conditions: I would rather see all validation errors in one place.

However if we want to programmatically consume these conditions or want to give our users this ability - then it will definitevely be easier to consume if we have conditions for each issue. But in this specific case I do not see a lot of value in this feature (but I'm happy to be proven wrong).

> I suggest we leave it as it is for now. I will add a comment and a ticket so we can follow up on this after a more thorough discussion upstream. wdyt?

Agreed

Reply:
The condition type needs to be unique in the list of conditions.

Type and reason are part of the API so once we GA, those can't be removed or changed.

My recommendation is generally to keep the number of types and reasons as small as possible based on what clients specifically require.

In this case, I'd start with whatever type we're already using, maybe even use the existing failure reason, and then put the validation errors concatenated in the message.

But +1 to capturing this in a separate issue.